dali/pipeline/operators/crop/bbox_crop.h

```diff
@@ -0,0 +1,261 @@
+// Copyright (c) 2017-2018, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_PIPELINE_OPERATORS_CROP_CROP_H_
+#define DALI_PIPELINE_OPERATORS_CROP_CROP_H_
+
+#include <vector>
+#include <utility>
+#include <algorithm>
+
+#include "dali/common.h"
+#include "dali/error_handling.h"
+#include "dali/pipeline/operators/common.h"
+#include "dali/pipeline/operators/operator.h"
+#include "dali/pipeline/operators/fused/crop_cast_permute.h"
+
+namespace dali {
+
+class BBoxCrop : public Operator<CPUBackend> {
+  static const unsigned int kAttempts = 100;
+  static const unsigned int kBboxSize = 4;
+
+ protected:
+  struct Bounds {
+    explicit Bounds(std::vector<float> &&bounds)
+        : min_(bounds[0]), max_(bounds[1]) {
+      DALI_ENFORCE(min_ >= 0, "Min should be at least 0.0. Received: " +
+                                  std::to_string(min_));
+      DALI_ENFORCE(bounds.size() == 2, "Bounds should be provided as 2 values");
+      DALI_ENFORCE(min_ <= max_, "Bounds should be provided as: [min, max]");
+    }
+
+    bool Contains(float k) const { return k >= min_ && k <= max_; }
+
+    const float min_, max_;
+  };
+
+  struct Rectangle {
+    explicit Rectangle(float left, float top, float right, float bottom)
+        : left_(left),
+          top_(top),
+          right_(right),
+          bottom_(bottom),
+          area_(right * bottom) {}
+
+    bool Contains(float x, float y) const {
+      return x >= left_ && (x - left_) <= right_ && y >= top_ &&
+             (y - top_) <= bottom_;
+    }
+
+    Rectangle ClampTo(const Rectangle &other) const {
+      const auto left = std::max(other.left_, left_);
+      const auto top = std::max(other.top_, top_);
+      const auto right = std::min(other.right_, right_);
+      const auto bottom = std::min(other.bottom_, bottom_);
+
+      return Rectangle(left, top, right, bottom);
+    }
+
+    float IntersectionOverUnion(const Rectangle &other) const {
+      const float intersection_area = ClampTo(other).area_;
+
+      return intersection_area /
+             static_cast<float>(area_ + other.area_ - intersection_area);
+    }
+
+    const float left_, top_, right_, bottom_, area_;
```

Comment:
Google style guide requires to use regular names for fields of a struct.

Reply:
accepted. Will apply it to other structs